### PR TITLE
[RFC][python 3.12] upgrade pytorch 

### PR DESCRIPTION
**Pull request by can-anyscale**
This PR upgrades pytorch to its latest version, so we can subsequently upgrade our tests to run with python 3.12. A few things:
- I skipped running horovod related tests on CI, but keep the code so teams can run them locally; horovod has been 1 year stale and is not compatible with a recent version of torch
- Upgrade CUDA to 12.1.1 to be compatible with the current version of torch + tensorflow

Test:
- CI

### CODE REVIEW

**Comment by can-anyscale** [comment]
request-for-comment versions since I skip horovod tests on CI with this upgrade, @matthewdeng, @justinvyu let me know what you think, thankks

**Comment by can-anyscale** [comment]
we still have horovod covered from the release tests i think, let me see if i can keep that

**Review by aslonnie — COMMENTED** [review]
yeah, probably train/tune team should review first.

File: rllib/core/learner/tests/test_learner_group.py [hunk 1]
```diff
@@ -148,6 +149,7 @@ def test_learner_group_build_from_algorithm_config(self):
         print(learner_group)
         learner_group.shutdown()
 
+    @pytest.mark.skip(reason="Does not work properly with latest torch 2.3.1 version")
```
Comment (can-anyscale) [inline rllib/core/learner/tests/test_learner_group.py:152]: CC: @sven1977 we'll fix this forward

**Review by justinvyu — APPROVED** [review]
@can-anyscale I listed out some alternatives [here](https://docs.google.com/document/d/1qyip2MvJ8KTCJm0IpyPNJlzuuRJQDB2i-2uqNkREgFs/edit#heading=h.wk4yjiurzq38).

I think the proposal in this PR makes sense.

**Comment by can-anyscale** [comment]
love the proposal, thanks for creating it!

**Comment by can-anyscale** [comment]
lol torch doesn't build for mac x86 anymore, so this is breaking our x86 macos build; i'm reverting

**Pull request merged**